lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3544,7 +3567,16 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
     // Note: we relax the restriction if the type we're casting to is a
     // non-final class because it's possible that we might have a subclass
     // that conforms to the protocol.
-    if (fromExistential && !toExistential) {
+    //
+    // Also, relax the restriction when toType is a bounded generic with
+    // an existential argument because it may have conditional protocol 
+    // conformances. e.g.
+    //
+    //  func encodable(_ value: Encodable) {
+    //    _ = value as! [String : Encodable]
+    //  }
```

Comment:
Hmm, I'm not sure this applies generally to all generic types, it seems to me that it's specific to the standard library collection types that we implement custom casting behaviour for.

For example, let's say `value` is storing a `[String: Int]`. What's really happening here is a two-step cast:

```
Encodable -> [String: Int] -> [String: Encodable]
```

Where the first step unwraps the existential, and the second step relies on the special casting behaviour where `Array`, `Set` and `Dictionary` can individually cast their elements.

It seems like rather than checking for existentials in the target type, we need to skip checking conditional requirements when we're checking the conformance of `Array`, `Set`, `Dictionary` to the protocol, as it's possible that an intermediate cast of the element type could cause the requirement to no longer be satisfied. You should be able to use `lookupConformance` to do that.

Reply:
Humm, I'll take a look at this :)